prow/genyaml/genyaml.go

```diff
@@ -0,0 +1,306 @@
+	if field.Names != nil {
+		return field.Names[0].Name
+	}
+	return field.Type.(*ast.Ident).Name
+}
+
+// genDocMap extracts the name of the field as it should appear in YAML format and returns the resultant string.
+func genDocMap(path string) (map[string]map[string]Comment, error) {
+	pkg, err := astFrom(path)
+	if err != nil {
+		return nil, errors.New("unable to generate AST documentation map")
+	}
+
+	dm := make(map[string]map[string]Comment)
+
+	for _, t := range pkg.Types {
+		if typeSpec, ok := t.Decl.Specs[0].(*ast.TypeSpec); ok {
+			if structType, ok := typeSpec.Type.(*ast.StructType); ok {
+
+				typeSpecName := typeSpec.Name.Name
+				dm[typeSpecName] = make(map[string]Comment)
+
+				for _, field := range structType.Fields.List {
+
+					if tagName := fieldName(field); tagName != "-" {
+						typeName := fieldType(field)
+						isObj := false
+
+						// Find leaf node.
+						ast.Inspect(field, func(n ast.Node) bool {
+							switch x := n.(type) {
+							case *ast.Ident:
+								typeName = x.Name
+								isObj = x.Obj != nil
+							}
+
+							return true
+						})
+
+						docString := fmtRawDoc(field.Doc.Text())
+						dm[typeSpecName][tagName] = Comment{typeName, isObj, docString}
+					}
+				}
+			}
+		}
+	}
+
+	return dm, nil
+}
+
+// extractTypeSpec updates the typeSpec for subsequent YAML keys for contextual comment decorating by type.
+func extractTypeSpec(parent *yaml3.Node, prevSibling *yaml3.Node, typeSpec string) string {
+	if parent.Kind == yaml3.MappingNode && prevSibling.Kind == yaml3.ScalarNode {
+		if subTypeSpec, ok := CommentMap[typeSpec][prevSibling.Value]; ok && subTypeSpec.IsObj {
+			return subTypeSpec.Type
+		}
+	}
+
+	return typeSpec
+}
+
+// injectComment reads a YAML node and injects a head comment based on its value and typeSpec.
+func injectComment(parent *yaml3.Node, typeSpec string) {
+	if v, ok := CommentMap[typeSpec][parent.Value]; ok {
+		parent.HeadComment = v.Doc
+	}
+
+	if parent.Content != nil {
+		for i, child := range parent.Content {
+			if i > 0 {
+				prevSibling := parent.Content[i-1]
+				typeSpec = extractTypeSpec(parent, prevSibling, typeSpec)
+			}
+
+			// Recurse to inject comments on nested YAML nodes.
+			injectComment(child, typeSpec)
+		}
+	}
+
+}
+
+// getType returns the type's name within its package for a defined type. For other (non-defined) types it returns the empty string.
+func getType(typ interface{}) string {
+	if t := reflect.TypeOf(typ); t.Kind() == reflect.Ptr {
+		return t.Elem().Name()
+	} else {
+		return t.Name()
+	}
+}
+
+// SetPath allow for setting of the global CommentMap via path specification to a `.go` file, returning a boolean indicating success.
+func SetPath(path string) bool {
+	commentMap, err := genDocMap(path)
+	if err != nil {
+		return false
+	}
+
+	CommentMap = commentMap
+	return true
+}
+
+// GenYaml generates a fully commented YAML snippet for a given plugin configuration.
+func GenYaml(config interface{}) string {
```

Comment:
nit: I think it would be more idiomatic to return `(string, error)` here and let the caller choose how to handle errors instead of logging them.